sdk/clientcore/core/src/main/java/io/clientcore/core/util/auth/ChallengeHandler.java

```diff
@@ -0,0 +1,29 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package io.clientcore.core.util.auth;
+
+import java.util.Arrays;
+
+/**
+ * Abstract class representing a challenge handler for authentication.
+ */
+public abstract class ChallengeHandler {
+
+    /**
+     * Handles the authentication challenge.
+     *
+     * @return The authorization header or token.
+     */
+    public abstract String handle();
```

Comment:
I believe we'll want a bit more depth here as some values are only available on the request or response and I don't think we'll want a pattern where the users of these types need to new-up an instance each network request.

Some thoughts on additional API changes:

- Proxies should return a `Proxy-Authenticate` header on a 407 response (`Proxy Authentication Required`) which should contain which auth mechanism to use. We should add an API to this abstract class (and maybe make this an interface?) which indicates whether the `ChallengeHandler` implementation is capable of authenticating the given challenge. This would also help the `CompositeChallengeHandler` as the method to handle authentication isn't performing double duty as authenticating and indicating authentication is possible.
- For handling authentication, I think it may be a better design to pass the `HttpRequest` being updated and the `HttpResponse` received with the challenge to be passed and this being a `void` method. This will allow `ChallengeHandler`s to be stateless, if their implementation chooses to do so, and will allow them to not need external handling to manage state. For example, in this commit `AuthorizationChallengeHandler` needs to manage some state for the `ChallengeHandler` being used to authenticate which will make it harder for external implementations to be integrated.

Reply:
updated the challenge handler API to be able to handle updating the request auth headers and reading response haders for auth challenges.